assertj-core/src/main/java/org/assertj/core/internal/TypeHolder.java

```diff
@@ -37,15 +40,16 @@
  */
 abstract class TypeHolder<T> {
 
-  private static final Comparator<Class<?>> DEFAULT_CLASS_COMPARATOR = ClassNameComparator.INSTANCE;
+  private static final Comparator<Pair<Class<?>, Class<?>>> DEFAULT_CLASS_COMPARATOR = new PairComparator<>(ClassNameComparator.INSTANCE,
+                                                                                                            ClassNameComparator.INSTANCE);
 
-  protected final Map<Class<?>, T> typeHolder;
+  protected final Map<Pair<Class<?>, Class<?>>, T> typeHolder;
```

Comment:
I feel it would be clearer to introduce a class that show the intention behind `Pair<Class<?>, Class<?>` which is a bit too generic, maybe `DualClass` ? still generic but at least save a pair of `<>`, is consistent with `DualValue` and instead of `left`, `right ` it could use `actual` and `expected` which is what we are comparing here

Reply:
renamed in 122d8af03bf46ff5716bf578d37b0cf6da4bf4aa